### PR TITLE
Switch user if failing to read /proc/<pid>/maps

### PR DESCRIPTION
**Pull request by nsavoire**
# Description

If open of `/proc/<pid>/maps` fails, then try to switch to same user as `/proc/<pid>/maps` for opening the file.

### CODE REVIEW

**Comment by nsavoire** [comment]
Current dependencies on/for this PR:
* main
  * **PR #189** <a href="https://app.graphite.dev/github/pr/DataDog/ddprof/189" target="_blank">Graphite</a>  👈
    * **PR #191** <a href="https://app.graphite.dev/github/pr/DataDog/ddprof/191" target="_blank">Graphite</a> 

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/DataDog/ddprof/189?utm_source=stack-comment).

File: src/user_override.cc
Comment (r1viollet) [inline src/user_override.cc:95]: Wow, thanks for spending some amount of time figuring this out ! That is a lot of operations. Interesting that the group is necessary.

File: test/dso-ut.cc
Comment (r1viollet) [inline test/dso-ut.cc:385]: thanks for adding this in !

**Review by r1viollet — APPROVED** [review]
LGTM

File: src/user_override.cc
Comment (nsavoire) [reply src/user_override.cc:26]: outcome is nice for returning a failure or a value, here it's pure RAII without error handling (failures are not relevant, it's a best effort), therefore there is no much benefit to using benefit.
It would had made sense to use outcome for `FileHolder` to return an error instead of an empty `unique_ptr`, but I prefer to introduce outcome in a later PR.

**Comment by nsavoire** [comment]
**Graphite Merge Job**

Current status: 🔄 Waiting for checks to pass

This pull request is currently waiting for CI and mergeability checks to pass before being merged.

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/DataDog/ddprof/189?utm_source=stack-comment).


Job Reference: SPsYsiNiTyhESOPX7N64

**Pull request merged**
